src/main/java/com/google/devtools/build/lib/exec/SymlinkTreeHelper.java

```diff
@@ -43,36 +42,35 @@
-  private final Path outputManifest;
   private final Path symlinkTreeRoot;
   private final String workspaceName;
 
   /**
    * Creates SymlinkTreeHelper instance. Can be used independently of SymlinkTreeAction.
    *
-   * @param inputManifest path to the input runfiles manifest
-   * @param outputManifest path to the output runfiles manifest
-   * @param symlinkTreeRoot path to the root of the symlink tree
-   * @param filesetTree true if this is a fileset symlink tree, false if this is a runfiles symlink
-   *     tree.
+   * @param inputManifest exec path to the input runfiles manifest
+   * @param symlinkTreeRoot the root of the symlink tree to be created
    * @param workspaceName the name of the workspace, used to create the workspace subdirectory
    */
-  public SymlinkTreeHelper(
-      Path inputManifest, Path outputManifest, Path symlinkTreeRoot, String workspaceName) {
-    this.inputManifest = ensureNonSnapshotting(inputManifest);
-    this.outputManifest = ensureNonSnapshotting(outputManifest);
-    this.symlinkTreeRoot = ensureNonSnapshotting(symlinkTreeRoot);
+  public SymlinkTreeHelper(Path inputManifest, Path symlinkTreeRoot, String workspaceName) {
+    // Do not indirect paths through overlay filesystems (such as the action filesystem or a
+    // snapshotting filesystem), for a few reasons:
+    // (1) we always want to create the symlinks on disk, even if the overlay filesystem creates
+    //     them in memory (at the time of writing, no action filesystem implementations do so, but
+    //     this may change in the future).
+    // (2) current action filesystem implementations are not a true overlay filesystem, so errors
+    //     might occur in an incremental build when the parent directory of a symlink exists on disk
+    //     but not in memory (see https://github.com/bazelbuild/bazel/issues/24867).
+    // (3) Changes made to a file referenced by a symlink tree should be reflected in the symlink
+    //     tree without having to rebuild. Therefore, if a snapshotting file system is used, we must
+    //     use the underlying non-snapshotting file system instead to create the symlink tree.
+    this.inputManifest = inputManifest.onUnderlyingFileSystem();
+    this.symlinkTreeRoot = symlinkTreeRoot.onUnderlyingFileSystem();
     this.workspaceName = workspaceName;
   }
 
-  private static Path ensureNonSnapshotting(Path path) {
-    // Changes made to a file referenced by a symlink tree should be reflected in the symlink tree
-    // without having to rebuild. Therefore, if a snapshotting file system is used, we must use the
-    // underlying non-snapshotting file system instead to create the symlink tree.
-    if (path.getFileSystem() instanceof SnapshottingFileSystem snapshottingFs) {
-      return snapshottingFs.getUnderlyingNonSnapshottingFileSystem().getPath(path.asFragment());
-    }
-    return path;
+  private Path getOutputManifest() {
+    return symlinkTreeRoot.getChild("MANIFEST");
```

Comment:
IMO the point of passing in `outputManifest` as an argument is that it avoids creating an additional point of truth for the name of the output manifest file.